CustomOp Inline Fusion

tianrengao opened the pull request:
Add Inline Fusion Support for Custom Op Autotuning
--------------------------------------------------

This PR extends PyTorch Inductor's custom op autotuning with inline fusion capabilities, enabling the winning decomposition to be inlined directly into the computation graph for fusion with surrounding operations.


### Usage

```python

def decompose_k_implementation(
    a: torch.Tensor, b: torch.Tensor, k_splits: int = 4
) -> torch.Tensor:
    """Matrix multiply with k-way decomposition."""
    ...

@torch.library.custom_op("my_lib::matmul_relu", mutates_args={})
def custom_matmul_relu_dk(
    a: torch.Tensor, b: torch.Tensor, k_splits: int 
) -> torch.Tensor:
    return torch.relu(decompose_k_implementation(a, b, k_splits))

register_custom_op_autotuning(
    custom_op=custom_matmul_relu_dk,
    configs=[
        CustomOpConfig(k_splits=2),
        CustomOpConfig(k_splits=4),
        CustomOpConfig(k_splits=8),
        CustomOpConfig(k_splits=32),
        CustomOpConfig(k_splits=64),
    ],
    name="decompose_k_autotuned",
    input_gen_fns={
        "a": lambda fake: torch.randn_like(fake, device='cuda'),
        "b": lambda fake: torch.randn_like(fake, device='cuda'),
    }
)
```

### How It Works
Enable optimizations from Inductor by inlining the best decomposition, allowing fusion with surrounding elementwise operations and other graph-level optimizations. This provide potentially better performance and memory efficiency.
During customop autotuning phase, we still benchmarks all CustomOpConfigs to find the fastest implementation. Then during inline fusion, inductor inline the decompositions into the main graph, converting the winning choice to individual ComputedBuffer IR nodes (fusable). At the end, Inductor automatically fuses inlined operations with surrounding elementwise ops (e.g., bias add, ReLU, scaling). Note that the winning choice must be a SubgraphChoiceCaller (decomposition-based) rather than an ExternKernelChoice for inlining to work. If the ExternKernelChoice is returned, no inline happens.


Performance Results
Benchmarked on matmul+relu workload with decompose-k fusion (H100 GPU, 15 test shapes):
<img width="782" height="377" alt="Screenshot 2025-11-04 at 12 43 11 AM" src="https://github.com/user-attachments/assets/22131d4c-a8ce-4f55-bdcd-ac758ddad8cd" />


Metric | Result
-- | --
Average Speedup vs ATen | 1.28x
Max Speedup vs ATen | 1.41x

<br class="Apple-interchange-newline">

The performance comparison are detailed in the below plots. We spot that on most use cases, the inline fusion gains better performance compared to aten baseline and the current torch.compile. 
<img width="4874" height="3545" alt="image" src="https://github.com/user-attachments/assets/190a1233-412f-4f34-84cd-9b7cb582f504" />

**Test**: `test_decompose_k_with_fusion` demonstrates decompose-k with inline fusion enabled.

--------------


### Integration to mm.py decomposeK with a flag enable_inline_subgraph_fusion=True in config (deprecated to avoid breaking async compilation. removed from the PR already)
FP32:
<img width="738" height="357" alt="Screenshot 2025-11-04 at 12 05 08 AM" src="https://github.com/user-attachments/assets/ee421d22-c426-42f2-8dcd-4dcc547d6219" />
FP16:
<img width="769" height="403" alt="Screenshot 2025-11-04 at 12 13 49 AM" src="https://github.com/user-attachments/assets/346d1ffc-15af-40b0-9378-cf9b297711c2" />

The TCF column represents torch compile fusion, which is close to custom_op decomposek. The difference might due to different candidate k values.


#### Usage:
Note: this only happens when we don't benchmark_epilogue_fusion, i.e., not using multi_template_buffer.

```python
# Define the matmul+relu function
    def matmul_relu(x, y):
        return torch.nn.functional.relu(torch.matmul(x, y))

    # Compile with inline subgraph fusion enabled
    @torch.compile
    def compiled_matmul_relu(x, y):
        return matmul_relu(x, y)

    # Reset dynamo to ensure clean compilation
    torch._dynamo.reset()

    with config.patch(
        {
            "max_autotune": True,
            # CRITICAL: These two flags enable inline subgraph fusion
            "benchmark_epilogue_fusion": False,  # Must be False for inline fusion!
            "enable_inline_subgraph_fusion": True,  # Enable inline fusion
        }
    ):
        # Compile and run
        result = compiled_matmul_relu(a, b)
        torch.cuda.synchronize()
```



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

PaulZhang12 commented:
Nice! This is awesome, curious why is there 1 case in which the perf was worse?

PaulZhang12 commented on torch/_inductor/select_algorithm.py:
```diff
@@ -2700,6 +2700,7 @@ def __call__(
         precompilation_timeout_seconds: int = 60 * 60,
         return_multi_template=False,
         best_config_future=None,
+        return_choice=False,
```
Fine for now, perhaps a refactor where we only return choice is better, as we can get node from choice

PaulZhang12 reviewed (approved):
LGTM! Probably wait until @eellison takes a look as well :). Also, does this automatically enable decomposeK fusions now if decomposeK wins with max_autotune_gemm? Or that is still yet to be implemented?

eellison reviewed (commented):
please rebase onto main since much of this code has already landed

tianrengao commented on torch/_inductor/kernel/custom_op.py:
```diff
@@ -0,0 +1,544 @@
+log = logging.getLogger(__name__)
+
+
+class CustomOpConfig:
```
Rebased it onto main and composed the inline fusion to mm.py when not using multiTemplateBuffer. Updated the benchmark result for Torch Compile with gemm max-autotune and inline fusion in the PR description.

Yea will make sure no landed code for review in the future.

tianrengao commented on torch/_inductor/config.py:
```diff
@@ -693,6 +701,11 @@ def use_autoheuristic(name: str) -> bool:
+# Enable inline fusion for subgraphs (custom_op, mm.py decomposek, etc.)
+# Only effective when benchmark_epilogue_fusion=False (immediate autotune path)
+# When True, inline the winning subgraph choice to IR nodes for epilogue fusion
+enable_inline_subgraph_fusion = False
```
shall we make this defualt or a flag?

eellison commented on torch/_inductor/kernel/custom_op.py:
```diff
@@ -318,20 +321,70 @@ def autotune_custom_op(
     )
 
+    # Apply inlining for fusion if winning_choice has graph; otherwise return result as-is(default impl)
+    if hasattr(winning_choice, "gm"):
```
We shouldn't be conditionally assigning attrs. can this be None if it's not set ?

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -1315,6 +1315,39 @@ def _to_dtype(x):
+        and inductor_config.enable_inline_subgraph_fusion
+    ):
+        result, winning_choice = autotune_select_algorithm(
+            name,
```
This interferes with async compilation. if you were to turn this on a do a benchmark run with  max-autotune, youd see a huge regresssion.

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -1315,6 +1315,39 @@ def _to_dtype(x):
+        )
+
+        # If choice has graph to lower, inline it; otherwise use result as-is(cublas mm)
+        if hasattr(winning_choice, "gm"):
```
can we do a better check then hasatt?

eellison commented on torch/_inductor/codegen/subgraph.py:
```diff
@@ -23,6 +23,69 @@
+    placeholder_idx = 0
+
+    # Process each node in the graph
+    for node in gm.graph.nodes:
```
is there a refactor we could do ?

eellison reviewed (commented):
The main reason we have MultiTemplateBuffer and deferred choice is so that we can overlap compilations of max-autotune. We shouldn't be turning this for our own mm implementations, at least if there are any non decomposition /extern choices.

I'm willing to approve this pr for user custom kernels, since it's unlikely they're spawning 10+ triton mm compilations, and we don't have the overlapping working there yet. It's a little unfortunate that this pr doesn't fix the mm decomposition case (without eagerly awaiting all async compilations). 

have a few, comments, first.

eellison reviewed (commented):
The main reason we have MultiTemplateBuffer and deferred choice is so that we can overlap compilations of max-autotune. We shouldn't be turning this for our own mm implementations, at least if there are any non decomposition /extern choices.

I'm willing to approve this pr for user custom kernels, since it's unlikely they're spawning 10+ triton mm compilations, and we don't have the overlapping working there yet. It's a little unfortunate that this pr doesn't fix the mm decomposition case (without eagerly awaiting all async compilations). 

have a few, comments, first.

eellison reviewed (commented):
The main reason we have MultiTemplateBuffer and deferred choice is so that we can overlap compilations of max-autotune. We shouldn't be turning this for our own mm implementations, at least if there are any non decomposition /extern choices.

I'm willing to approve this pr for user custom kernels, since it's unlikely they're spawning 10+ triton mm compilations, and we don't have the overlapping working there yet. It's a little unfortunate that this pr doesn't fix the mm decomposition case (without eagerly awaiting all async compilations). 

have a few, comments, first.

tianrengao committed:
resolve comments
Summary:
- refactor inline_subgraph function.
- remove mm.py inline fusion logic to avoid breaking parallel async compilation.
- remove unused flags.
- remove mlp tests as it is not useful and customop autotuning features have been covered by other tests.
